### ISSUE TITLE
KeywordRecognizer | Python | macOS: stop_recognition_async() does not work

### ISSUE THREAD

**Issue by CtrlZvi** [Azure-Samples/cognitive-services-speech-sdk#1045]
**Describe the bug**
Calling `stop_recognition_async()` after `recognize_once_async()` does not appear to work correctly.
1. It takes about 10 seconds to do whatever it is it does.
2. The `canceled` callback do not run as a result of invoking it.
3. `get()` on the `ResultFuture` returned by `recognize_once_async()` blocks forever.

**To Reproduce**
1. Run the below proof of concept
```python3
import azure.cognitiveservices.speech as speechsdk
import time


def speech_recognize_keyword_locally_from_microphone():
    """runs keyword spotting locally, with direct access to the result audio"""

    # Creates an instance of a keyword recognition model. Update this to
    # point to the location of your keyword recognition model.
    model = speechsdk.KeywordRecognitionModel("YourKeywordRecognitionModelFile.table")

    # The phrase your keyword recognition model triggers on.
    keyword = "YourKeyword"

    # Create a local keyword recognizer with the default microphone device for input.
    keyword_recognizer = speechsdk.KeywordRecognizer()

    done = False

    def recognized_cb(evt):
        # Only a keyword phrase is recognized. The result cannot be 'NoMatch'
        # and there is no timeout. The recognizer runs until a keyword phrase
        # is detected or recognition is canceled (by stop_recognition_async()
        # or due to the end of an input file or stream).
        result = evt.result
        if result.reason == speechsdk.ResultReason.RecognizedKeyword:
            print("RECOGNIZED KEYWORD: {}".format(result.text))
        nonlocal done
        done = True

    def canceled_cb(evt):
        result = evt.result
        if result.reason == speechsdk.ResultReason.Canceled:
            print("CANCELED: {}".format(result.cancellation_details.reason))
        nonlocal done
        done = True

    # Connect callbacks to the events fired by the keyword recognizer.
    keyword_recognizer.recognized.connect(recognized_cb)
    keyword_recognizer.canceled.connect(canceled_cb)

    # Start keyword recognition.
    result_future = keyword_recognizer.recognize_once_async(model)
    print(
        'Say something starting with "{}" followed by whatever you want...'.format(
            keyword
        )
    )
    # If active keyword recognition needs to be stopped before results, it can be done with
    #
    stop_future = keyword_recognizer.stop_recognition_async()
    print(f"{time.monotonic()} Stopping...")
    stopped = stop_future.get()
    print(f"{time.monotonic()} Stopped")
    result = result_future.get()
    print(f"{time.monotonic()} Done")

    # Read result audio (incl. the keyword).
    if result.reason == speechsdk.ResultReason.RecognizedKeyword:
        time.sleep(2)  # give some time so the stream is filled
        result_stream = speechsdk.AudioDataStream(result)
        result_stream.detach_input()  # stop any more data from input getting to the stream

        save_future = result_stream.save_to_wav_file_async(
            "AudioFromRecognizedKeyword.wav"
        )
        print("Saving file...")
        saved = save_future.get()
    else:
        result_stream = speechsdk.AudioDataStream(result)
        result_stream.detach_input()


if __name__ == "__main__":
    speech_recognize_keyword_locally_from_microphone()
```

**Expected behavior**
The `canceled` callbacks are called, `get()` on the future returned by `recognize_once_async()` returns a result indicating cancellation, and `get()` on the future returned by `stop_recognition_async()` returns almost immediately.

**Version of the Cognitive Services Speech SDK**
1.16.0

**Platform, Operating System, and Programming Language**
 - OS: macOS Big Sur 11.2.3.
 - Hardware: x64
 - Programming language: Python
 - Browser: N/A

**Additional context**
I'm assuming that `get()` on the future returned by `recognize_once_async()` should still return after stopping because the documentation says "A future that is fulfilled once recognition has been initialized." "Initialized" doesn't seem to require successful recognition, which leads me to infer that it should still return on cancellation.

**Comment by CtrlZvi** [comment]
It appears that the call to `recognize_once_async()` did not, in fact cancel anything, despite returning. If I speak the hotword, after cancellation, the program will recognize the hotword and return a successful result. If I speek the hotword before the cancellation finishes (during the 10 second window), then when the cancellation completes, the cancellation callback _is_ invoked, but, again the hotword is also recognized and a successful recognition is returned.

**Comment by CtrlZvi** [comment]
As an interesting counterpoint, though, if I wrap the keyword recognition into a coroutine and run the recognition on a separate thread, then the cancellation works just fine (see below sample). Seems like there might be a threading issue?

If I use the result the in the original coroutine to convert the result to an `AudioDataStream` and then `detach_input()`. If I do that in the thread that's getting the recognition result, everything works fine (I assume that's intended, but it makes pushing the the `get()` of the recognition to a new thread a bit more complicated).

```python3
import asyncio
from asyncio import CancelledError
from pathlib import Path
from os import PathLike
from typing import Union
import time

import azure.cognitiveservices.speech as speechsdk


async def _recognize(
    model_path: Union[PathLike, str, bytes]
) -> speechsdk.KeywordRecognitionResult:
    model = speechsdk.KeywordRecognitionModel(str(model_path))
    keyword_recognizer = speechsdk.KeywordRecognizer()

    def recognized_callback(event):
        result = event.result
        print(f"{time.monotonic()}: Recognized: {result.reason}")
        if result.reason == speechsdk.ResultReason.RecognizedKeyword:
            print(f"{time.monotonic()}: RECOGNIZED KEYWORD: {result.text}")

    def canceled_callback(event):
        result = event.result
        print(f"{time.monotonic()}: Canceled: {result.reason}")
        if result.reason == speechsdk.ResultReason.Canceled:
            print(f"{time.monotonic()}: CANCELED: {result.cancellation_details.reason}")

    keyword_recognizer.recognized.connect(recognized_callback)
    keyword_recognizer.canceled.connect(canceled_callback)

    def _get_result():
        result = keyword_recognizer.recognize_once_async(model).get()
        # This works fine
        print(f"{time.monotonic()}: Creating result stream")
        result_stream = speechsdk.AudioDataStream(result)
        print(f"{time.monotonic()}: Result stream created")
        result_stream.detach_input()
        print(f"{time.monotonic()}: Input detached")
        
        return result

    try:
        print(f"{time.monotonic()}: Recognizing")
        return await asyncio.to_thread(_get_result)
    except CancelledError:
        print(f"{time.monotonic()}: Cancelling")
        keyword_recognizer.stop_recognition_async().get()
        print(f"{time.monotonic()}: Cancelled")
        raise


async def main():
    recognition_task = asyncio.create_task(
        _recognize(Path("YourKeywordRecognitionModelFile.table"))
    )
    try:
        print(f"{time.monotonic()}: Sleeping")
        await asyncio.sleep(2)
        print(f"{time.monotonic()}: Cancelling")
        recognition_task.cancel()
        print(f"{time.monotonic()}: Awaiting")
        result = await recognition_task
    except:
        print(f"{time.monotonic()}: Caught exception")
        recognition_task.cancel()
        print(f"{time.monotonic()}: Recognition canceled")
        return
    print(f"{time.monotonic()}: Await successful")

    print(result)
    if result.reason == speechsdk.ResultReason.RecognizedKeyword:
        print(f"{time.monotonic()}: RECOGNIZED KEYWORD: {result.text}")
        # This would stall
        # print(f"{time.monotonic()}: Creating result stream")
        # result_stream = speechsdk.AudioDataStream(result)
        # print(f"{time.monotonic()}: Result stream created")
        # result_stream.detach_input()
        # print(f"{time.monotonic()}: Input detached")
    print(f"{time.monotonic()}: All done")


if __name__ == "__main__":
    print(f"{time.monotonic()}: Starting loop")
    asyncio.run(main())
    print(f"{time.monotonic()}: Loop complete")
```

**Comment by chschrae** [comment]
Thanks for reporting this issue. You've provided everything we need to look into this and we're tracking the investigation on the SDK team's backlog. To keep the open issues list fresh, we'll close this issue while it's prioritized and developed against and report back here once we have an update. Please subscribe to the issue to get the latest and please open a new issue (that can freely reference this one) if it was closed prematurely or there's new/different information.

Also feel free to add any information you gather to this issue if you discover more. It sounds like you are doing a great investigation and all info is greatly appreciated.

**Comment by oscholz** [comment]
reopening and labeling it as accepted so @CtrlZvi gets notified when we resolve this after we fix this bug.

**Comment by pankopon** [comment]
Internal work item ref. 3475645.

**Comment by glecaros** [comment]
This has been fixed and it will be available in the next release. Will close this issue when the next release is published

**Comment by pankopon** [comment]
Closed as the Speech SDK 1.24.0 has been released.
